packages/flutter_tools/lib/src/build_system/targets/macos.dart

```diff
@@ -83,111 +127,151 @@ class UnpackMacOS extends Target {
+      manifestPath: environment.projectDir.childFile('pubspec.yaml').path,
+      packagesPath: environment.projectDir.childFile('.packages').path,
     );
-  }
-}
-
-/// Build all of the artifacts for a debug macOS application.
-class DebugMacOSApplication extends Target {
-  const DebugMacOSApplication();
-
-  @override
-  Future<void> build(List<File> inputFiles, Environment environment) async {
-    final File sourceFile = environment.buildDir.childFile('app.dill');
-    final File destinationFile = environment.buildDir
-        .childDirectory('flutter_assets')
-        .childFile('kernel_blob.bin');
-    if (!destinationFile.parent.existsSync()) {
-      destinationFile.parent.createSync(recursive: true);
+    if (result != 0) {
+      throw Exception('Failed to create asset bundle: $result');
+    }
+    // Limit number of open files to avoid running out of file descriptors.
+    try {
+      final Pool pool = Pool(64);
+      await Future.wait<void>(
+        assetBundle.entries.entries.map<Future<void>>((MapEntry<String, DevFSContent> entry) async {
+          final PoolResource resource = await pool.request();
+          try {
+            final File file = fs.file(fs.path.join(assetDirectory.path, entry.key));
+            file.parent.createSync(recursive: true);
+            await file.writeAsBytes(await entry.value.contentsAsBytes());
+          } finally {
+            resource.release();
+          }
+        }));
+    } catch (err, st){
+      throw Exception('Failed to copy assets: $st');
+    }
+    // Copy dill file.
+    try {
+      final File sourceFile = environment.buildDir.childFile('app.dill');
+      sourceFile.copySync(assetDirectory.childFile('kernel_blob.bin').path);
+    } catch (err) {
+      throw Exception('Failed to copy app.dill: $err');
+    }
+    // Copy precompiled runtimes.
+    try {
+      final String vmSnapshotData = artifacts.getArtifactPath(Artifact.vmSnapshotData,
+          platform: TargetPlatform.darwin_x64, mode: BuildMode.debug);
+      final String isolateSnapshotData = artifacts.getArtifactPath(Artifact.isolateSnapshotData,
+          platform: TargetPlatform.darwin_x64, mode: BuildMode.debug);
+      fs.file(vmSnapshotData).copySync(
+          assetDirectory.childFile('vm_snapshot_data').path);
+      fs.file(isolateSnapshotData).copySync(
+          assetDirectory.childFile('isolate_snapshot_data').path);
+    } catch (err) {
+      throw Exception('Failed to copy precompiled runtimes: $err');
     }
-    sourceFile.copySync(destinationFile.path);
   }
 
   @override
   List<Target> get dependencies => const <Target>[
-    FlutterPlugins(),
-    UnpackMacOS(),
     KernelSnapshot(),
-    CopyAssets(),
-    DebugMacOSPodInstall(),
+    DebugMacOSFramework(),
+    UnpackMacOS(),
   ];
 
   @override
   List<Source> get inputs => const <Source>[
-    Source.pattern('{BUILD_DIR}/app.dill')
+    Source.behavior(MacOSAssetBehavior()),
+    Source.pattern('{BUILD_DIR}/app.dill'),
```

Comment:
Taking the patch for a test drive locally, I noticed that this ended up as an output (I didn't dig into the full derivation of why). Should the final output list maybe filter out anything the `.dart_tool` directory as being an intermediate?

Reply:
Added code to filter out .dart_tool